added button tests and support for btn-default as a default style.

thinkingmedia opened the pull request:
This PR improves buttons.

- closes #143 as button widget now assigns `btn-default` if it's missing.
- ButtonWidget now checks for both `primary` and `btn-primary` classname patterns in class attribute.
- Adds unit test coverage for ButtonWidget
- Fixes `btn-default` for `Form->submit()` which didn't use ButtonWidget.

I added a public static method to ButtonWidget which is called from FormHelper. So that there is only one function that handles class styles for buttons. I'm not sure if a helper should call a static on a widget. Seems dirty to me. Is this okay for now?

ADmad commented on tests/TestCase/View/Widget/ButtonWidgetTest.php:
```diff
@@ -0,0 +1,197 @@
+    /**
+     * @var StringTemplate
+     */
+    private $templates;
```
Change to `public`.

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -160,8 +161,8 @@ public function submit($caption = null, array $options = [])
         }
-
         $options = Hash::merge($options, ['class' => $class]);
+        $options['class'] = ButtonWidget::applyStyle($options['class']);
```
Don't like this. You can add the method `applyStyle()` to `OptionsAwareTrait`. That trait is already used in the helpers and widgets.

thinkingmedia replied:
OptionsAwareTrait is part of CakePHP.. That method sets the specific styles for bootstrap.

thinkingmedia replied:
Maybe I could add a new trait to this project?

ADmad replied:
@thinkingmedia I am referring to this one https://github.com/FriendsOfCake/bootstrap-ui/blob/master/src/View/Helper/OptionsAwareTrait.php :)

thinkingmedia replied:
WTF! I must be getting old. I thought that was in cake. lol

ADmad replied:
There's one with same classname in cake too but under different namespace for different purpose :)

thinkingmedia commented:
I had to change this in `composer.json` because I wanted to use the `pre-commit` hook for git.

```
    "require-dev": {
        "phpunit/phpunit": "4.1.*",
        "squizlabs/php_codesniffer": "2.*",
        "cakephp/cakephp-codesniffer": "dev-master"
    },
```

Now the hook is working.

thinkingmedia commented:
This PR has been update

- Adds more class functions to the OptionsAwareTrait
- Adds unit test for OptionsAwareTrait
- Button/Form updated to use OptionsAwareTrait for setting button styles.

ADmad commented on composer.json:
```diff
@@ -20,7 +20,9 @@
     "require-dev": {
-        "phpunit/phpunit": "4.1.*"
+        "phpunit/phpunit": "4.1.*",
+        "squizlabs/php_codesniffer": "2.*",
```
This is not needed. `cakephp/cakephp-condesniffer` already has it as a dependency.

ADmad commented on composer.json:
```diff
@@ -20,7 +20,9 @@
-        "phpunit/phpunit": "4.1.*"
+        "phpunit/phpunit": "4.1.*",
+        "squizlabs/php_codesniffer": "2.*",
+        "cakephp/cakephp-codesniffer": "dev-master"
```
Better to use `~2.0`.

ADmad commented on src/View/Helper/OptionsAwareTrait.php:
```diff
@@ -5,7 +5,96 @@
+     *
+     * @var array
+     */
+    public $ButtonStyles = [
```
`$buttonStyle` instead of `$ButtonStyles`.

ADmad replied:
Also `$buttonClasses` would be better.

ADmad commented on src/View/Helper/OptionsAwareTrait.php:
```diff
@@ -5,7 +5,96 @@
+     *
+     * @var array
+     */
+    public $ButtonStyleAliases = [
```
`$ButtonStyleAliases` => `$buttonClassAliases`.

ADmad commented on src/View/Helper/OptionsAwareTrait.php:
```diff
@@ -5,7 +5,96 @@
+     * @param array $data An array of HTML attributes and options.
+     * @return array An array of HTML attributes and options.
+     */
+    public function applyButtonStyles(array $data)
```
`applyButtonClasses`

ADmad replied:
This method name change still needs to be addressed :)

ADmad commented on src/View/Widget/ButtonWidget.php:
```diff
@@ -10,15 +10,6 @@ class ButtonWidget extends \Cake\View\Widget\ButtonWidget
 
     use OptionsAwareTrait;
 
-    protected $_styles = [
```
Hmm.. Removing the property is not BC.

ADmad replied:
Mark it deprecated instead.

thinkingmedia commented:
updated PR

- renamed public properties
- removed squizlabs/php_codesniffer
- update git hook script (copies from cakephp so it now does a php lint)

thinkingmedia commented:
updated PR

- restored property and made it deprecated

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -2,6 +2,7 @@
 
 namespace BootstrapUI\View\Helper;
 
+use BootstrapUI\View\Widget\ButtonWidget;
```
This needs to be removed.

thinkingmedia commented:
updated PR

- removed BudgetWidget use from FormHelper

thinkingmedia commented:
updated PR

- renamed method to applyButtonClasses

thinkingmedia commented:
boy that was suppose to be an easy change. lol

thinkingmedia commented:
bump

ADmad commented:
@jadb Thoughts?
